homeassistant/helpers/device_registry.py

```diff
@@ -201,40 +205,40 @@ def _update_device(self, old_device: DeviceEntry, new_device: DeviceEntry) -> No
         _remove_device_from_index(devices_index, old_device)
         _add_device_to_index(devices_index, new_device)
 
-    def _clear_index(self):
+    def _clear_index(self) -> None:
         """Clear the index."""
         self._devices_index = {
             REGISTERED_DEVICE: {IDX_IDENTIFIERS: {}, IDX_CONNECTIONS: {}},
             DELETED_DEVICE: {IDX_IDENTIFIERS: {}, IDX_CONNECTIONS: {}},
         }
 
-    def _rebuild_index(self):
+    def _rebuild_index(self) -> None:
         """Create the index after loading devices."""
         self._clear_index()
         for device in self.devices.values():
             _add_device_to_index(self._devices_index[REGISTERED_DEVICE], device)
-        for device in self.deleted_devices.values():
-            _add_device_to_index(self._devices_index[DELETED_DEVICE], device)
+        for deleted_device in self.deleted_devices.values():
+            _add_device_to_index(self._devices_index[DELETED_DEVICE], deleted_device)
 
     @callback
     def async_get_or_create(
         self,
         *,
-        config_entry_id,
-        connections=None,
-        identifiers=None,
-        manufacturer=UNDEFINED,
-        model=UNDEFINED,
-        name=UNDEFINED,
-        default_manufacturer=UNDEFINED,
-        default_model=UNDEFINED,
-        default_name=UNDEFINED,
-        sw_version=UNDEFINED,
-        entry_type=UNDEFINED,
-        via_device=None,
+        config_entry_id: str,
+        connections: Optional[set] = None,
```

Comment:
Shouldn't connections and identifiers be a set of two item tuples of strings? We type it to that on the DeviceEntry.

Reply:
I haven't looked deeper, but I think they could. Did not touch that part though, because it would likely require changes to parts of the code that already had some type hints (albeit too lax/with room for improvement; untyped collections), such as `_normalize_connections`. I wanted to keep the changes to the minimum to reach/match the goal described in the PR title: to have complete type hints. Improving/making them stricter could be done in future PRs.

Reply:
Sounds good!